pageserver/src/http/routes.rs

```diff
@@ -73,23 +68,45 @@ async fn status_handler(request: Request<Body>) -> Result<Response<Body>, ApiErr
     )?)
 }
 
-async fn branch_create_handler(mut request: Request<Body>) -> Result<Response<Body>, ApiError> {
-    let request_data: BranchCreateRequest = json_request(&mut request).await?;
+async fn timeline_create_handler(mut request: Request<Body>) -> Result<Response<Body>, ApiError> {
+    let tenant_id: ZTenantId = parse_request_param(&request, "tenant_id")?;
+    let request_data: TimelineCreateRequest = json_request(&mut request).await?;
 
-    check_permission(&request, Some(request_data.tenant_id))?;
+    check_permission(&request, Some(tenant_id))?;
 
-    let response_data = tokio::task::spawn_blocking(move || {
-        let _enter = info_span!("/branch_create", name = %request_data.name, tenant = %request_data.tenant_id, startpoint=%request_data.start_point).entered();
-        branches::create_branch(
+    let new_timeline_info = tokio::task::spawn_blocking(move || {
+        let _enter = info_span!("/timeline_create", tenant = %tenant_id, new_timeline = ?request_data.new_timeline_id, lsn=?request_data.ancestor_start_lsn).entered();
+        timelines::create_timeline(
             get_config(&request),
-            &request_data.name,
-            &request_data.start_point,
-            &request_data.tenant_id,
+            tenant_id,
+            request_data.new_timeline_id.map(ZTimelineId::from),
+            request_data.ancestor_timeline_id.map(ZTimelineId::from),
+            request_data.ancestor_start_lsn,
         )
     })
     .await
     .map_err(ApiError::from_err)??;
-    Ok(json_response(StatusCode::CREATED, response_data)?)
+
+    Ok(match new_timeline_info {
+        Some(info) => json_response(StatusCode::CREATED, TimelineInfoResponse::from(info))?,
+        None => json_response(StatusCode::NOT_MODIFIED, ())?,
```

Comment:
I would rather return 400 with timeline already exists message, WDYT?

Reply:
I think it is a client error to create already existing tenant, I'm ok with either 400 or 409. I can imagine using NOT_MODIFIED status when you asked server to update something, but server state already meets the update criteria. Here we're trying to create new resource, rather than modifying already existing one. Usually for 400 response on some e-commerse websites there is not only a message but also a code message. E g `{"code": "invalid_json", "msg": "Invalid json"}`. Where msg can even be translated to multiple languages. Another point is when there will be multiple kinds of errors we probably won't use different 4xx codes for them, or even when there may be a more than one meaning of some particular error code, e g different types of conflicts. So we will have to distinguish between them anyway. 

I think it is simpler to use an umbrella 400 for that, but do not insist and for current simple case any option is fine.

Reply:
> I think it is a client error to create already existing tenant

I have one general argument against it, motivating me to return non-error status: if it's an error, the cluster has to defend itself against it by either listing clusters/tenants before, to ensure that we're not recreating anything or by calling verification methods after errors.
At its current model, where any failure would be retried, the cluster has to protect itself from these errors by having an extra call.
Now, when we've split tenant and timeline creation, this means 4 requests in this model, vs 2 requests in a "non-error" one, apparently that [might be important](https://github.com/zenithdb/console/pull/782#discussion_r822860234).

----

So, the "error" approach generally requires more defensive request to be made since is not treated as an error really, still hardly see the benefits myself.

But I actually like the idea of `200 OK {"code": "invalid_json", "msg": "Invalid json"}` approach as more flexible indeed, so maybe it's not worth touching the current odd status now: two repos have to be changed for almost nothing.
But sure feel free to adjust the code yourself later 🙂

Reply:
> Now, when we've split tenant and timeline creation, this means 4 requests in this model, vs 2 requests in a "non-error" one, apparently that https://github.com/zenithdb/console/pull/782#discussion_r822860234.

I do not get it why two requests are needed, client just retries until either 201 or 400 with already exists error code. There is a saying in python community: “it’s easier to ask for forgiveness than permission”, I think it might be applicable here

I've found a [thread](https://stackoverflow.com/q/3825990/6378300) on SO with many possible options and 409 being the most upvoted one.

Note that safekeepers recently gained the ability to create timelines on them via http api, so I think we should come to a solution which then will be used everywhere. 

Another example with an error is our console which uses 400 during sign up if email is already taken